### ISSUE TITLE
ShouldBeEquivalentTo case insensitive property match

### ISSUE THREAD

**Issue by nick-randal** [fluentassertions/fluentassertions#254]
How can I use ShouldBeEquivalentTo case insensitive property match.  i.e. A.PersonId and B.PersonID.

**Comment by dennisdoomen** [comment]
Not out of the box. But what you can do is implement [IMemberMatchingRule](https://github.com/dennisdoomen/fluentassertions/blob/develop/FluentAssertions.Core/Equivalency/IMemberMatchingRule.cs) and pass that to the `Using()` call on the `options` parameter of `ShouldBeEquivalentTo`.

**Comment by deadlydog** [comment]
Any progress on this? I feel like it's a common enough scenario that it would be nice to have an out-of-the-box solution for it.

**Comment by JSkimming** [comment]
I've just needed the same thing. If it helps anyone, here's my implementation that ignores case when matching properties.

* [IMemberMatchingRule](https://github.com/fluentassertions/fluentassertions/blob/5.9.0/Src/FluentAssertions/Equivalency/IMemberMatchingRule.cs). 👈 _(This link won't break)_

```csharp
public class CaseInsensitiveMatch : IMemberMatchingRule
{
    public SelectedMemberInfo Match(
        SelectedMemberInfo expectedMember,
        object subject,
        string memberPath,
        IEquivalencyAssertionOptions config)
    {
        List<PropertyInfo> properties =
            subject.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance)
                .Where(pi => pi.Name.Equals(expectedMember.Name, StringComparison.OrdinalIgnoreCase))
                .ToList();

        PropertyInfo property = properties.Count > 1
            ? properties.SingleOrDefault(p => p.PropertyType == expectedMember.MemberType)
            : properties.SingleOrDefault();

        SelectedMemberInfo match = SelectedMemberInfo.Create(property);
        return match;
    }
}
```

**Comment by deadlydog** [comment]
Cool, thanks for the code @JSkimming :) 

I also went with my own implementation that I'll share here. I chose to just not use the `BeEquivalentTo` method and just used this bit I found on Stack Overflow.

```csharp
actualStringList.Should().Equal(expectedStringList, (s1, s2) => string.Equals(s1, s2, StringComparison.OrdinalIgnoreCase));
```

**Comment by rynkevich** [comment]
I've been thinking on this for some time and can't come up with a case where this could be really useful, as usually you won't compare two object of different types. You create an `expected` object of the same type as `actual` and compare them. If you can't create it, you make your own class with the same member names that you have in `actual`. @nick-randal @JSkimming @deadlydog, can you please share your use cases for this feature?

However, if this feature is to be implemented, I'd like to have an ability to perform case-insensitive matches not just considering letters in lower and upper case equivalent, but also ignoring the differences in naming conventions like `snake_case` or `CamelCase`. This may look like
```
foo.Should().BeEquivalentTo(bar, options => options.IgnoringCaseInMemberNames(ignoreUnderscores: true));
```
Since cases like `SCREAMING_SNAKE_CASE`, `camelCase` would be covered by default and `kebab-case` could not be used to create a valid C# identifier, the only thing required to support such comparison is to ignore underscores during the comparison. However, there's a corner case with identifiers like `_foo`, `__Bar`, `Baz__` etc.

**Comment by JSkimming** [comment]
@rynkevich the use case is when I want to compare two objects where I either can't or don't want to change one to be consistent with the other.

For instance, if I'm using a third-party library to retrieve some data, and then save it to an Entity Framework DB Context.

The third part library has an object with the properties `FirstName` and `Lastname`, but my existing EF models use `Firstname` and `Lastname`.

I could:

1. Change my EF models to use the inconsistent naming convention of the third party library.
2. I could make FluentAssertions compare properties ignoring case.

I prefer option 2 because I don't want to propagate the inconsistent naming convention to my models.

I can also imagine a scenario where the third party library names things with underscores (I've found this some generated third party libraries that are built on frameworks that use the underscore convention). I agree that it would also be useful. But the simple case adding an option to enable case insensitive appears, on the surface, to be a relatively simple option to allow.

**Comment by rynkevich** [comment]
@JSkimming, thanks for the example. I guess, it's all about just providing a more convenient API for specific usage scenarios. The question is don't we try to make FluentAssertions perform more and more complex mapping logic? This isn't AutoMapper or something.

@dennisdoomen, could you share your thoughts on that?

Speaking about third party libraries with strange conventions, an example that comes to my mind is [PayPal SDK](https://github.com/paypal/PayPal-NET-SDK/tree/develop/Source/SDK/Api). They use underscores for property names and that could be where this feature could be applied, if you write some abstration.

**Comment by dennisdoomen** [comment]
I think it's a fair question to ask for something like that, but considering its edgy scenarios, I don't think it _really_ belongs in the library. So I would not reject a PR that tries to implement it, but I also don't feel the need to do this myself.

**Comment by rynkevich** [comment]
@nick-randal, here's my two cents:
- This feature probably should be more than just case insensitive match for properties. To support other scenarios also addressed by other features in library we should check `IEquivalencyAssertionOptions` to see which members have to be included in match. There's an example [here](https://github.com/fluentassertions/fluentassertions/blob/c192ca014ae920c96f76a747b72bf110c31ba153/Src/FluentAssertions/Equivalency/Matching/MustMatchByNameRule.cs).
- A new configuration method could be added directly to [EquivalencyAssertionOptions](https://github.com/fluentassertions/fluentassertions/blob/c192ca014ae920c96f76a747b72bf110c31ba153/Src/FluentAssertions/Equivalency/EquivalencyAssertionOptions.cs).

**Comment by dennisdoomen** [comment]
Yep. the direction is fine and the remarks from @rynkevich are completely valid.

**Comment by dennisdoomen** [comment]
This is solved by #1742
